docs/en/integrations/data-ingestion/data-formats/json.md

```diff
@@ -0,0 +1,576 @@
+
+```sql
+CREATE TABLE events
+(
+    `data` String
+)
+ENGINE = MergeTree
+ORDER BY ()
+```
+
+Now we can load data from the file into this table using [JSONAsString](/docs/en/interfaces/formats.md/#jsonasstring) format to keep JSON objects instead of parsing them:
+
+```sql
+INSERT INTO events (data)
+FROM INFILE 'custom.json'
+FORMAT JSONAsString
+```
+
+And we can use [JSON functions](/docs/en/sql-reference/functions/json-functions.md) to query saved objects:
+
+```sql
+SELECT
+    JSONExtractString(data, 'type') AS type,
+    data
+FROM events;
+
+┌─type───┬─data─────────────────────────────────────────────────┐
+│ person │ {"name": "Joe", "age": 99, "type": "person"}         │
+│ post   │ {"url": "/my.post.MD", "hits": 1263, "type": "post"} │
+│ log    │ {"message": "Warning on disk usage", "type": "log"}  │
+└────────┴──────────────────────────────────────────────────────┘
+```
+
+Consider using [JSONAsObject](/docs/en/interfaces/schema-inference.md/#json-as-object) together with a new [JSON data type](/docs/en/guides/developer/working-with-json/json-semi-structured.md/#json-object-type) to store and process JSON in tables in a more efficient way. Note that JSONAsString works perfectly fine in cases we have JSON object-per-line formatted files (usually used with `JSONEachRow` format).
+
+## Data types detection when importing JSON data
+
+ClickHouse does some magic to guess the best types while importing JSON data. We can use a `DESCRIBE` clause to check which types were defined:
+
+```sql
+DESCRIBE TABLE file('list.json', JSONEachRow);
+
+┌─name──┬─type─────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
+│ path  │ Nullable(String) │              │                    │         │                  │                │
+│ month │ Nullable(Date)   │              │                    │         │                  │                │
+│ hits  │ Nullable(Int64)  │              │                    │         │                  │                │
+└───────┴──────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
+```
+
+This allows quickly creating tables from JSON files:
+
+```sql
+CREATE TABLE new_table
+ENGINE = MergeTree
+ORDER BY tuple() AS
+SELECT *
+FROM file('list.json', JSONEachRow)
+```
+
+Detected types will be used for this table:
+
+```sql
+DESCRIBE TABLE new_table;
+
+┌─name──┬─type─────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
+│ path  │ Nullable(String) │              │                    │         │                  │                │
+│ month │ Nullable(Date)   │              │                    │         │                  │                │
+│ hits  │ Nullable(Int64)  │              │                    │         │                  │                │
+└───────┴──────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
+```
```

Comment:
Worth a nested example here as it uses Tuples and Nested type...and you can see the full type hierarchy?

Reply:
automatic inference won't work here, but I've added a section/file to demonstrate how this can happen "JSON objects with nested objects".